### PR TITLE
Driver Permissions Policy

### PR DESCRIPTION
**Pull request by ssands1**
### Pull Request Overview

This pull request implements a permissions policy which blocks apps from accessing drivers to which they lack explicit access as set in the app's header. If the app lacks the requisite permissions, the board pretends the driver doesn't exist.

The TBF header now includes a new `tipe = 5` which is a series of bytes which form a bitmap with each bit corresponding to one driver, as determined by the order of driver numbers in `capsules/src/driver.rs`.

See companion pull request [elf2tab PR #9](https://github.com/tock/elf2tab/pull/9). Tockloader PR coming soon.

### Testing Strategy

To test this PR alone, I specify which permissions I want for a given app by hard-coding elf2tab to write a specific u64 in all TBF headers. To test different permissions, I currently change the hard-coded value `DEFAULT_PERMS` in [elf2tab/header.rs](https://github.com/tock/elf2tab/pull/9/files), but a future PR to tockloader will allow the user to set specific permissions from the command line (coming soon).

### TODO or Help Wanted

TODO:

- [X] Create [PR for elf2tab](https://github.com/tock/elf2tab/pull/9) to work with this feature
- [ ] Create PR for tockloader to work with this feature
- [ ] Update the relevant files in `/docs`

Help Wanted: 

There are two main downsides that I see so far:

1. tockloader must know the mapping of drivers to bits so when I type `tockloader add-permission my_app LED` (possible in a coming PR), tockloader knows which bit corresponds to the LED. This means another list of drivers to maintain in tockloader.

2. elf2tab must know how many bytes to leave blank as permission bits in each TAB. This means a constant at the top of elf2tab/header.rs that needs to be incremented each time a driver is added.

Any thoughts on a better way to do this would be appreciated.

Also: Thoughts on what drivers an app should have access to by default? All? None? Specific drivers?

### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [X] Ran `make formatall`.

### CODE REVIEW

File: boards/imix/src/main.rs [hunk 1]
```diff
@@ -446,6 +461,7 @@ pub unsafe fn reset_handler() {
     rf233.start();
 
     imix.pconsole.start();
+    imix.permissions.start();
```
Comment (alevy) [inline boards/imix/src/main.rs]: You don't _have_ to have a `start` function, and it isn't currently doing anything, so it's fine to just exclude it. Some other capsules have `start` functions because they actually need to initialize something after the rest of the board has been configured

File: capsules/src/permissions.rs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+//! Simple capsule that checks if a permission is set to true or false
+
+#![forbid(unsafe_code)]
```
Comment (alevy) [inline capsules/src/permissions.rs]: `forbid(unsafe_code)` is already covered in the crate's `lib.rs`, so you don't need to repeat it here.

File: capsules/src/permissions.rs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    }
+
+    pub fn start(&self) {}
+
+    pub fn check(&self, permissions: u64, driver_num: usize) -> bool {
+        if let Some(bit) = crate::driver::get_permission_bit(driver_num) {
+            permissions & (1 << bit) == (1 << bit)
+        } else {
+            // driver not found
+            false
```
Comment (alevy) [inline capsules/src/permissions.rs]: This implies that drivers have to be whitelisted. I.e. if they aren't mentioned in the policy, they are default off. Not a bad policy, but should be noted somewhere (and maybe configurable?)

File: kernel/src/platform/mod.rs [hunk 4]
```diff
@@ -13,6 +13,13 @@ pub trait Platform {
     fn with_driver<F, R>(&self, driver_num: usize, f: F) -> R
     where
         F: FnOnce(Option<&Driver>) -> R;
+
+    fn with_driver_permissions<F, R>(&self, _permissions: u64, driver_num: usize, f: F) -> R
```
Comment (alevy) [inline kernel/src/platform/mod.rs]: add a comment explaining what this method does, and that it's default implementation is a no-op (i.e. it just blindly calls `with_driver`)

File: kernel/src/sched.rs [hunk 5]
```diff
@@ -331,19 +349,32 @@ impl Kernel {
...
                                     allow_size,
                                 }) => {
-                                    let res = platform.with_driver(driver_number, |driver| {
-                                        match driver {
-                                            Some(d) => {
-                                                match process.allow(allow_address, allow_size) {
-                                                    Ok(oslice) => {
-                                                        d.allow(appid, subdriver_number, oslice)
+                                    let permissions: u64 = if let Some(header) =
+                                        tbfheader::parse_and_validate_tbf_header(
```
Comment (alevy) [inline kernel/src/sched.rs]: This seems like it should be done in `process.rs` when the process is loaded for the first time, and cached in the `Process` struct, so you don't need to parse and validate on every system call.

Comment (ssands1) [reply kernel/src/sched.rs]: Good call -- I moved it to the ProcessType trait -- I think this is a much better place for it.

**Comment by lschuermann** [comment]
I think this PR introduces a really cool feature!

However, in it's current state this would introduce another list of drivers to maintain - in the kernel and eventually also tockloader. I don't know if there's a better way to approach this though.

In addition to that, this seems to be limited to 64 drivers due to the use of `u64`, right? As we're already at 30, it's not unlikely to assume that we're exceeding this number in the future? Would it be possible to make the size of the TBF header field dynamic?

**Comment by ssands1** [comment]
Hi @lschuermann --
> This seems to be limited to 64 drivers due to the use of `u64`, right? As we're already at 30, it's not unlikely to assume that we're exceeding this number in the future? Would it be possible to make the size of the TBF header field dynamic?
I agree, it's not the best way. The problem is, elf2tab has to know how many bytes to leave free for permissions when it compiles TABs. One solution that I'm thinking could work to make the number of bytes expandable is if I pass an argument to tell elf2tab how many bytes to leave free (and have a default number of bytes if no argument is passed, say 64 or 128 or 256). However, then tockloader and tock would need to know how many bytes to parse permissions from. I'm currently rewriting the above PR to use the `TbfHeaderTLV` type which includes a `length` field; tockloader and tock can read the length and parse the next `length` bytes as permissions. Using the `TbfHeaderTLV` block will also allow me to not need to create a `TbfHeaderV3` the way I thought I had to for backwards compatibility. Thoughts on this solution?

>  In it's current state this would introduce another list of drivers to maintain - in the kernel and eventually also tockloader. I don't know if there's a better way to approach this though.
Again, the problem is that if I want to be able to tell tockloader which permissions I need, then tockloader has to know which bit corresponds to which permission. I could either store this in tockloader as it is now, or I could store this in the `TbfHeader` and have tockloader parse out the mappings from each app, but I think that way would be less desirable. I'm not sure of any better way to do it than it is now.

**Comment by ssands1** [comment]
I'm running into two problems when trying to make the number of bytes for permissions dynamic:

1. I can't dynamically allocate space in the kernel, as that's not allowed. Therefore, the kernel must know how many bytes to parse at compile time. 

2. I can't pass the number of drivers to elf2tab as an argument because the makefile doesn't know the number of drivers either.

My current implementation uses an array of u8s, but `const PERM_LENGTH` is required in both the tock kernel and in elf2tab and will need to be updated each time a driver is added.

Thoughts?

**Comment by ssands1** [comment]
I've made expandable the number of permission bytes to be read. The only non-ideal side effect is: How does elf2tab know how many permission bytes to leave room for in the TBF header? It would need to know the number of drivers. Thus, I had to add a constant `NUM_DRIVERS` in elf2tab/header.rs that'll need to be incremented each time a driver is added.

File: capsules/src/driver.rs [hunk 6]
```diff
@@ -4,6 +4,10 @@ use enum_primitive::enum_from_primitive;
 enum_from_primitive! {
 #[derive(Debug, PartialEq)]
 // syscall driver numbers
+// when adding a driver, make sure to:
+//   - add the driver to `get_permissions_bit()` below
+//   - add the driver to `self.permission_bits` in tockloader/tockloader/tbfh.py
+//   - increment `NUM_DRIVERS` in elf2tab/header.rs
```
Comment (ssands1) [inline capsules/src/driver.rs]: Is there a better way?

**Comment by phil-levis** [comment]
Is per-driver the right level of control? I.e., are there cases when drivers should be allowed, but only partially? I would rather have one control mechanism than multiple ones. For example, if we decide we need to be able restrict which ADC pins a process can read, or how often, then we might want to express more than just driver numbers.

But let's suppose that we want per-driver control. It seems like there are two separate technical problems, with a third one that entwines them:
  1) How does a process binary specify which drivers it is allowed to use?
  2) How does the kernel check whether a running process can access a driver?
  3) How does the kernel map 1 to 2 when it loads a process?

I'd argue that for 1), a list of driver IDs in the header would be best.  It's perhaps not concise, but the number of drivers accessed will be bounded by code length, so processes which access lots of drivers will have a lot of code.

For 2), can we get away without runtime checks? Processes cannot execute out of RAM. Could we analyze the binary such that we can verify r0 on every instance of svc is an allowed value?

**Comment by alevy** [comment]
> 1), a list of driver IDs in the header would be best. It's perhaps not concise, but the number of drivers accessed will be bounded by code length, so processes which access lots of drivers will have a lot of code.

To reiterate @phil-levis's point, a few folks had raised this as an alternative approach in the weekly call on Friday.

The basic sense is that if we're allowing access based on drivers, we should do so using the already existing convention of [system call numbers](https://github.com/tock/tock/blob/master/doc/syscalls/README.md#capsule-provided-drivers) rather than adding another convention of mapping bit indexes to a subset of driver numbers.

In principle using a bitmask could be fine for this, except the system call number space is HUGE... E.g. the sensor drivers are in the range of 0x60000-0x6ffff, which would require half a million bits to represent as-is.

In other words, using a bit mask without another mapping just isn't a good fit for anything except the ["Basic" system calls](https://github.com/tock/tock/blob/master/doc/syscalls/README.md#base) (which start at `0x0`).

Instead, as @phil-levis suggested, an alternative approach would be to have a TLV that includes a list of system call numbers the process is allowed to access, (e.g. `[0x0, 0x1, 0x2, 0x60000]` for `[Alarm, Console, Led, Temperature]`). Then, each time the binary is loaded by the kernel, the _kernel_ would convert this list into some convenient format for enforcing (e.g. a bitmask), that can be platform specific.

@ssands1 give our brief chat about your changes to the TLV format, I actually don't think you're too far off from this design.

File: capsules/src/driver.rs [hunk 7]
```diff
@@ -37,3 +41,44 @@ pub enum NUM {
...
+            NUM::USB_USER => 9,
+            NUM::I2C_MASTER_SLAVE => 10,
+            NUM::BLE_ADVERTISING => 11,
+            NUM::RNG => 12,
+            NUM::CRC => 13,
+            NUM::I2C_MASTER => 14,
+            NUM::APP_FLASH => 15,
+            NUM::NVM_STORAGE => 16,
+            NUM::SD_CARD => 17,
+            NUM::TEMPERATURE => 18,
```
Comment (ssands1) [inline capsules/src/driver.rs]: @alevy With regards to a mapping of bits to driver numbers being huge because, for example, the driver number for temperature is `0x60000`, this is where the additional mapping takes place. As I have it, `0x60000` maps to `1 << 18`. That being said, it doesn't mean that Phil's suggestion might not be a better way -- just that I had already created the mapping that prevents the need for half a million bits.

**Comment by phil-levis** [comment]
Amit and I talked a bit about this proposal at CPS week. I think there's a tension between two problems:
  1. driver numbers, while useful, are clearly not the finest granularity at which we want to manage process capabilities, but
  1. we have no idea what the finer granularity might look like.

So while this is a good step in the right direction, it's not the final word. On the other hand, who knows how long it will take to figure out the final word: that might be a complex research project by itself. Blocking an obvious improvement out of concern for a hypothetical better solution is bad process.

I suggest that we define some semantics about how this TLV and future TLVs operate, and leave these future TLVs as an open question. Specifically:
  1. If a process image doesn't have a permissions TLV, then it has access to all system calls (backwards compatibility, keep the simple case simple). The kernel can optionally decide to refuse to load such processes.
  1. The permissions TLV whitelists which drivers a process can use, enumerated as their driver IDs. 
  1. Each driver listed in the permissions TLV may have its own permissions TLV. The format of this permissions TLV is driver-specific. If there is no driver-specific TLV, the process has access to all system calls. The driver-specific TLV may limit which system calls (or which parameters) are allowed.

To correctly support this last bullet, we'll need an API/trait by which a TLV is passed to and parsed by a driver, then applied on every system call, hopefully in the main system call dispatch (so a programmer doesn't have the change to forget it).

Thoughts?

**Comment by bradjc** [comment]
Closed due to age, but this PR is very promising and reasonably close to mergable. If someone picks this back up we would love to see this in Tock (probably after v1.4). Also, tracking issue in #1350.

**Comment by ssands1** [comment]
@bradjc I'd be happy to pick this back up; I'll just need some guidance with what needs fixing.
